src/shared_utilities/utils.jl

```diff
@@ -438,72 +362,145 @@ end
+property of the `affect!` property of the `DiscreteCallback`. For example:
 
-Important: The times in `saveat` must be times the simulation is
-evaluated at for this function to work.
+`cb = NonInterpSavingCallback(start_date, stop_date, callback_period)`
+`saved_values = cb.affect!.saved_values`
+`saved_times = saved_values.t` - Vector of DateTimes
+`saved_cache = saved_values.saveval` - A vector of caches
 
-Note that unlike SciMLBase's SavingCallback, this version does not
-interpolate if a time in saveat is not a multiple of our timestep. This
-function also doesn't work with adaptive timestepping.
+Note: This method constructs a callback that uses `ITime` for scheduling, and should only
+be used with simulations that are run with `ITime`.
 """
-function NonInterpSavingCallback(saved_values, saveat::Vector{FT}) where {FT}
-    # This assumes that saveat contains multiples of the timestep
-    cond = condition(saveat)
-    saveiter = 0
-    affect! = SavingAffect(saved_values, saveat, saveiter)
-
-    SciMLBase.DiscreteCallback(
-        cond,
-        affect!;
-        initialize = saving_initialize,
-        save_positions = (false, false),
+function NonInterpSavingCallback(
+    start_date::Dates.DateTime,
+    stop_date::Dates.DateTime,
+    callback_period::Dates.Period;
+    first_save_date = start_date,
+)
+    first_save_during_init = first_save_date == start_date
+    callback_start =
+        first_save_during_init ? first_save_date :
+        first_save_date - callback_period
+    n_saves = length(first_save_date:callback_period:stop_date)
+    sv = (;
+        t = Vector{Union{Dates.DateTime, Nothing}}(nothing, n_saves),
+        saveval = Vector{Any}(undef, n_saves),
+    )
+    return NonInterpSavingCallback(
+        sv,
+        callback_period;
+        t0 = start_date,
+        callback_start,
+        init_saving = first_save_during_init,
     )
 end
 
 """
-    condition(saveat)
+    NonInterpSavingCallback(t0, tf, callback_period; t_first_save = t0)
 
-This function returns a function with the type signature expected by
-`SciMLBase.DiscreteCallback`, and determines whether `affect!` gets
-called in the callback. This implementation simply checks if the current time
-is contained in the list of affect times used for the callback.
+Constructs a `DiscreteCallback` which saves the time and cache `p` at `t_first_save`
+and every `callback_period` after. Times are saved as the same type as `t0`,
+and the saved times and caches are each saved in a vector. They are stored in a `NamedTuple`,
+which is the `saved_values` property of the `affect!` property of the `DiscreteCallback`.
+For example:
+
+`cb = NonInterpSavingCallback(t0, tf, callback_period)`
+`saved_values = cb.affect!.saved_values`
+`saved_times = saved_values.t` - Vector of typeof(t0)
+`saved_cache = saved_values.saveval` - A vector of caches
+
+Note: This method constructs a callback that uses `ITime` for scheduling, and should only
```

Comment:
we use ITime everywhere but a small subset of experiments? or did we convert to ITime everywhere but not the simulation struct in the last 2-3 files?

can we assume we use ITime for the simulation time?